doc/mgmt/gnmi/SONiC_GNOI_Server_Interface_Design.md

```diff
@@ -0,0 +1,632 @@
+1.	The first part is type, it can be SONIC, DPU, CABLE, CPLD and ASIC, we might support other types like CPU microcode.
+2.	The second part is device name, it can be DPU0 for DPU, and it can be Ethernet0 or ASIC0-Ethernet1 for cable. It’s possible to have embedded device name, for example, DPU0/Ethernet0.
+3.	The last part is default for now, we will use this part for future extension.
+
+<img src="images/filename_gnoi.svg" alt="filename-gnoi" width="800px"/>
+
+We will use the same GNOI API call to activate existing DPU firmware and reboot DPU by setting activate to True.
+
+```
+// Package defines a single package file to be placed on the target.
+message Package {
+  // Destination path and filename of the package.
+  string filename = 1;
+  // Version of the package. (vendor internal name)
+  string version = 4;
+  // Indicates that the package should be made active after receipt on
+  // the device. For system image packages, the new image is expected to
+  // be active after a reboot.
+  bool activate = 5;
+  // Details for the device to download the package from a remote location.
+  common.RemoteDownload remote_download = 6;
+}
+
+// SetPackageRequest will place the package onto the target and optionally mark
+// it as the next bootable image. The initial message must be a package
+// message containing the filename and information about the file. Following the
+// initial message the contents are then streamed in maximum 64k chunks. The
+// final message must be a hash message contains the hash of the file contents.
+message SetPackageRequest {
+  oneof request {
+    Package package = 1;
+    bytes contents = 2;
+    types.HashType hash = 3;    // Verification hash of data.
+  }
+}
+
+message SetPackageResponse {
+}
+```
+Note: The OS.Activate API is not suitable for our use case, because it does not have a field to specify the subcomponent that we want to activate.
+
+* GNOI System.Reboot
+We can use System.Reboot API to reboot one or more DPU, and the RebootRequest has a subcomponents field that allows us to specify which DPU we want to reboot.
+```
+message RebootRequest {
+  RebootMethod method = 1;
+  // Delay in nanoseconds before issuing reboot.
+  uint64 delay = 2;
+  // Informational reason for the reboot.
+  string message = 3;
+  // Optional sub-components to reboot.
+  repeated types.Path subcomponents = 4;
+  // Force reboot if sanity checks fail. (ex. uncommited configuration)
+  bool force = 5;
+}
+message Path {
+  string origin = 2;                              // Label to disambiguate path.
+  repeated PathElem elem = 3;                     // Elements of the path.
+}
+message PathElem {
+  string name = 1;                    // The name of the element in the path.
+  map<string, string> key = 2;        // Map of key (attribute) name to value.
+}
+```
+* GNMI Get
+Read current DPU firmware. If it's equal to target version, then upgrade was successful. 
+
+### 5.2. Upgrade Cable Firmware
```

Comment:
Do you want to explain cable and dual tor?

Reply:
Since these upgrades are similar to the SONiC image upgrade already described in detail, I chose to highlight just the standard SONiC upgrade scenario